proposals/NNNN-staticbigint.md

```diff
@@ -0,0 +1,136 @@
+
+Swift-evolution thread: [Pitch](https://forums.swift.org/t/staticbigint/54545)
+
+## Motivation
+
+There are two compiler-intrinsic protocols for integer literals.
+
+```swift
+public protocol _ExpressibleByBuiltinIntegerLiteral {
+    init(_builtinIntegerLiteral value: Builtin.IntLiteral)
+}
+
+public protocol ExpressibleByIntegerLiteral {
+    associatedtype IntegerLiteralType: _ExpressibleByBuiltinIntegerLiteral
+    init(integerLiteral value: IntegerLiteralType)
+}
+```
+
+- All integer (and floating-point) types in the standard library conform to both protocols.
+- Types outside of the standard library can only conform to the second protocol.
+- Therefore, the associated `IntegerLiteralType` must be a standard library type.
+
+For example, if larger fixed-width integers (such as `UInt256`) were added to the [Swift Numerics][] package, they would currently have to use smaller literals (such as `UInt64`).
+
+```swift
+let value: UInt256 = 0x1_0000_0000_0000_0000
+//                   ^
+// error: integer literal '18446744073709551616' overflows when stored into 'UInt256'
+```
+
+## Proposed solution
+
+Swift Numerics could (perhaps conditionally) use `StaticBigInt` as an associated type.
+
+```swift
+extension UInt256: ExpressibleByIntegerLiteral {
+
+#if compiler(>=9999) && COMPILATION_CONDITION
+    public typealias IntegerLiteralType = StaticBigInt
+#else
+    public typealias IntegerLiteralType = UInt64
+#endif
+
+    public init(integerLiteral value: IntegerLiteralType) {
+        precondition(
+            value.signum() >= 0 && value.bitWidth <= 257,
+            "integer literal overflows when stored into 'UInt256'"
+        )
+        // Copy the elements of `value.words` into this instance.
+        // Avoid numeric APIs that may trigger infinite recursion.
+    }
+}
+```
+
+Overflow would be a runtime error, unless [compile-time][] evaluation can be used in the future.
+
+## Detailed design
+
+`StaticBigInt` is an *immutable* arbitrary-precision signed integer. It can't conform to any [numeric protocols][], but it does implement some numeric APIs: `signum()`, `bitWidth`, and `words`.
+
+```swift
+public struct StaticBigInt: Sendable, _ExpressibleByBuiltinIntegerLiteral {
+    public init(_builtinIntegerLiteral value: Builtin.IntLiteral)
+}
+
+// `Self` Literals
+extension StaticBigInt: ExpressibleByIntegerLiteral {
+    public init(integerLiteral value: Self)
+    public static prefix func + (_ rhs: Self) -> Self
+}
+
+// Numeric APIs
+extension StaticBigInt {
+    public func signum() -> Int
+    public var bitWidth: Int { get }
+    public var words: Words { get }
+}
+
+// Collection APIs
+extension StaticBigInt {
+    public struct Words: RandomAccessCollection, Sendable {
+        public typealias Element = UInt
+        public typealias Index = Int
+    }
+}
```

Comment:
Should there be a way to construct an arbitrary `ExpressibleByIntegerLiteral` type from a `StaticBigInt`?